### DISCUSSION TITLE
Defining _MB_CAPABLE leads to some inconsistencies with some ctype functions

### DISCUSSION THREAD

**Discussion by A-Shehab** [picolibc/picolibc#872] [Q&A] [answered]
Hi All,

I've encountered an issue while testing the **iswalnum()** function with wide character inputs.
 According to the C99 standard **7.25.2.1.1** The iswalnum function;
 **iswalnum()** should test for any wide character for which **iswalpha()** or **iswdigit()** is true.

**Issue:**

When I test **iswalnum()** using wide characters for example with hexadecimal values in the range **0x0660 to 0x0669** (Arabic-Indic digits), the function returns non-zero if _MB_CAPABLE is defined as it calls the **POSIX-1.2008 iswalnum_l()** version That categorizes those hexadecimal values under the CAT_Nd (Numeric_Decimal) category in Unicode which is correct according to https://www.compart.com/en/unicode/category/Nd
which explains why iswalnum() identifies them as alphanumeric in this case.

However, the **iswdigit()** function, which follows the C99 standard strictly, returns non-zero only for numbers 0 through 9, which is also expected according to C99.

**Additional Observations:**

This behavior isn't limited to **iswalnum()** but also occurs in **iswblank, iswcntrl, iswprint, and tolower**. When **_MB_CAPABLE** is defined, these functions call their _l versions, could lead to inconsistencies with the C99 standard.

Test 1:
```
#include <wctype.h>
#include <wchar.h>
#include <limits.h>
#include <stdio.h>

int main(void)
{
    wint_t wc = 1632; //0x660 which is '0' in arabic

    if((!!iswalnum(wc) == (!!iswalpha(wc) || !!iswdigit(wc))) == 0)
    {
        printf("Test failed\n");
        printf("iswalnum = %d", iswalnum(wc));
        printf("iswalpha = %d", iswalpha(wc));
        printf("iswdigit = %d", iswdigit(wc));
    }
    else
    {
        printf("Test Passed\n");
        printf("iswalnum = %d\n", iswalnum(wc));
        printf("iswalpha = %d\n", iswalpha(wc));
        printf("iswdigit = %d\n", iswdigit(wc));
    }

}
```
Output:
```
Test failed
iswalnum = 1
iswalpha = 0
iswdigit = 0
```
The output shows that this hex value is considered as alpha-numeric but at the same time its not alphabetical or numerical.

according to this: https://en.cppreference.com/w/cpp/string/wide/iswdigit
**iswdigit()** should return non-zero only for decimal values from 0-9

I'm not sure if using the POSIX "_l" functions wrapped inside the C99 functions is 100% ok and not contradicting with any of the standards or the "_l" should be used as a standalone functions not wrapped inside the C standard function. 
For consistency with the C99 standard, iswalnum() should ideally match the behavior of iswalpha() and iswdigit() without additional modifications based on Unicode categories when _MB_CAPABLE is set.

Would it be possible to consider an adjustment that aligns the function behavior with C99, even under _MB_CAPABLE?

Thank you,

**Comment by keith-packard** [comment]
Whoa. I think you found a conflict between POSIX and C about the definition of `iswdigit` here. According to POSIX, that's a bug in the POSIX spec. But the C spec is not consistent, which makes the POSIX error understandable.

C says that `iswdigit` (and `isdigit`) only match the ASCII digits 0-9 (I checked both C17 and C23):

> The **iswdigit** function tests for any wide character that corresponds to a decimal-digit character (as defined in 5.2.1).

Section 5.2.1 says that decimal digits are those found in the *basic character set* and do not include any locale-specific values from the extended set:

> the 10 decimal digits:
> `0 1 2 3 4 4 5 6 7 8 9`

The rest of the wctype functions all add language like: 

> or any wide character that is one of a locale-specific set of alphabetic wide characters

This allows locales to include characters outside of the _basic character set_ in these tests. The ctype macros have similar wording, and `isdigit` matches the above definition.

On the other hand, POSIX says that `iswdigit` is locale-specific and matches any digit in the current locale:

> The _iswdigit_() (and _iswdigit_l_()) functions shall test whether wc is a wide-character code representing a character of class **digit** in the current locale (or in the locale represented by locale) respectively.

However, POSIX also explains the function again in the *__RETURN VALUE__* section:

> The *iswdigit*() (and *iswdigit_l*())  functions shall return non-zero if *wc* is a decimal digit wide-character code; otherwise, they shall return 0.

So, POSIX is internally inconsistent here.

POSIX also says that any mis-match between this spec and the ISO C spec are it's problem:

> For *iswdigit*(): The functionality described on this reference page is aligned with the ISO C standard. Any conflict between the requirements described here and the ISO C standard is unintentional. This volume of POSIX.1-2024 defers to the ISO C standard.

So, I think we can safely interpret this according to the C spec and ignore the first POSIX definition. That would mean leaving `iswdigit` and `iswdigit_l` alone.

Looking to `iswalnum` and `iswalnum_l`, we don't find wording equivalent to the C spec anywhere, instead it's all locale-specific. However, now that we've interpreted `iswdigit` to match the C spec, we have effectively reverse-engineered what the locale class of **digit** means -- it always means just the digits `0..9` in the basic character set. So, we should define `iswalnum` to be how C defines it (`iswalpha(c) || iswdigit(c)`), and then define `iswalnum_l` to be `iswalpha_l(c,l) || iswdigit_l(c, l)`.

I'm not sure I *like* this answer, but it is at least the most consistent interpretation I can build given the ambiguous language in the POSIX spec, and the fact that POSIX explicitly says that conflicts must be resolved in favor of the C spec.

**Reply by A-Shehab** [reply]
@keith-packard, Thank you for detailed answer regarding this conflict.
investigating more ctype functions i noticed in **iswprint_l()** that it returns 1 with **U2028** & **U2029** unicodes which are line & paragraph separators respectively and these are categorized as **cntrl** chars not **print** chars according to https://www.open-std.org/JTC1/SC35/WG5/docs/30112d10.pdf pg.35 and in the table here https://en.cppreference.com/w/cpp/string/wide/iswprint there is no value for **iswprint()** that returns non-zero also retruns non-zero with **iswcntrl()** in addition to that, in POSIX;

> The iswprint( ) and iswprint_l( ) functions shall test whether wc is a wide-character code
> representing a character of class print in the current locale of the process, or in the locale
> represented by locale

I was thinking of opening a Pull Request with this change that could fix this issue if you think according to the references mentioned this is correct.

```
int
iswprint_l (wint_t c, struct __locale_t *locale)
{
  (void) locale;
#ifdef _MB_CAPABLE
  c = _jp2uc_l (c, locale);
  enum category cat = category (c);
  return cat != -1
      && cat != CAT_Cc && cat != CAT_Cf
      && cat != CAT_Cs // Surrogate
>>      && cat != CAT_Zl && cat != CAT_Zp // Line/Paragraph Separators
      
      ;
#else
  return c < (wint_t)0x100 ? isprint (c) : 0;
#endif /* _MB_CAPABLE */
}
```

**Reply by keith-packard** [reply]
I just wrote a simple test that compares glibc and picolibc results for all of the wctype functions. We've got some rather significant differences, and every one I checked showed glibc with correct answers and picolibc with incorrect answers. To fix this, I'll start by updating the unicode database used by picolibc, then go evaluate how glibc uses that information compared with picolibc.

**Comment by keith-packard** [comment, marked as answer]
Ok, I couldn't fix the existing implementation; that only used the Unicode 'category' field to try and classify code points. I've replaced it with code that computes all of the classification data in a python program and then simply checks those at runtime. That python program uses the same semantics as glibc, so the results should match that implementation. Checkout #878

**Reply by A-Shehab** [reply]
Thank you for your time.
